### DISCUSSION TITLE
Enhancement: Feature to display previously pulled library

### DISCUSSION THREAD

**Discussion by Sailfish** [audiamus/BookLibConnect#75] [Ideas]
It would be useful to be able to redisplay the Library from the latest previously downloaded library and be able to select from that display to schedule a conversion. For people with relatively a small number of titles, it's no big deal but for those with a large number of titles (e.g., ~1,400 titles) it takes awhile. Even something simple like a run switch to signify to use latest downloaded list would be fine.

**Comment by audiamus** [comment]
I am not quite sure what you are envisaging.

You are aware that all columns are sortable? Sorting by date in decreasing order will list your latest purchases at the top.

**Reply by Sailfish** [reply]
![BLC-Run-Results](https://user-images.githubusercontent.com/110050/195676135-c066c809-95f7-473c-9fa5-2e6fd9c6ddec.jpg)

The attached image is of a recent run and took BLC about 40secs to download. The bottom pane titled 'Library for "Sailfish and region "us"' is, of course the list of titles from which I can choose to select title(s) in which I want BLC to transcode. However, my normal method is to only perform the transcoding on, at most, a couple of titles at a time. Then, after a varying amount of time later, I decide I wish to transcode another title or two that are already in the downloaded BLC library from the last time I ran the program.

What I would prefer is to have is some means to inform BLC (runtime switch, perhaps?) to not re-download the library again but, instead, to bring up the last version that had been downloaded before. Then, allow one to pick another set of titles to transcode from that list.

If this was possible, it would save time from having to wait for the Audible library title list download, thus saving Audible server load in addition to allowing me to start the process more quickly.

**Reply by audiamus** [reply]
Well, unless you click the Resync button, Book Lib Connect only downloads the meta data for your latest additions to the library, based on the time stamp of your previous access to the server. It usually takes a few seconds only. Check the log file, it should tell you how long.

What may take significantly longer is to create the big multi-column list with all your books when this window is first opened. The data is taken from the local database on your machine. The individual columns of this list are filled from various database tables which must be combined, observing the relations between them. There might be some potential for optimization here but that needs to be investigated carefully. Currently, most operations are carried out by the SQLite search engine. And that can be rather slow, despite using database indexes all over the place.

**Reply by Sailfish** [reply]
Checking the log file, the timestamped entries below appear to be the section where the construction of the library table starts and ends. Yes, it looks like that part took about 1.5 minutes.

Okay, as I mentioned above, it's not a big issue to me but since, on average, I purchase 2-3 titles a week, the reformatting time will continue to increase over time. It sounds like even without resync, you reconstruct the title catalog display from the SQLite db each time which would mean you'd have to save the constructed catalog display in order to be able to skip the SQLite formatting process. In a previous discussion, we were able to determine that Using Windows Select All (Ctr+A) and Copy (Ctr+C) on the formatted catalog display then Paste (Ctr+V) that into Excel would save the list in a Excel spreadsheet. Would saving that copied text file afford any speed improvement versus traversing the SQLite db?
//
11:07:59.511 0001 [MainForm.initLibraryAsync] <<< settings: auto refresh=True, auto update=True
11:07:59.511 0001 [MainForm.init] <<< 
11:07:59.522 0001 [MainForm.OnShown] <<< 
11:08:30.957 0006 [BookLibrary.GetBooks] from DB, #books=1385
11:08:30.957 0006 [BookLibrary.GetBooks] <<< ProfileId { AccountId = 1, Region = us }
11:08:30.958 0001 [BookLibDGVControl.setDataSource] #books=1385 (deliv type filtered)
11:08:30.961 0001 [BookLibDGVControl.resetDataSource] #books=1385 (filtered)
11:08:31.020 0001 [BookLibForm.bookLibdgvControl1_BookSelectionChanged] #books=1
11:08:31.074 0001 [BookLibForm.bookLibdgvControl1_BookSelectionChanged] #books=0
11:08:31.081 0001 [BookLibForm.bookLibdgvControl1_BookSelectionChanged] #books=1
11:08:31.095 0006 [BookLibrary.CheckUpdateFilesAndState] >>> 
11:08:32.240 0006 [BookLibrary.CheckUpdateFilesAndState] <<< 
11:08:32.240 0001 [BookLibForm.OnLoad] <<< 
11:09:13.433 0001 [BookLibForm.bookLibdgvControl1_BookSelectionChanged] #books=1
11:09:23.307 0001 [BookLibDGVControl.afterDownloadSelectionChanged] download selected: #books=1, #parts=2
11:09:23.310 0001 [BookLibForm.bookLibdgvControl1_BookDownloadSelectionChanged] #books=1
11:09:23.311 0001 [ConvertDGVControl.bookLibForm_BookDownloadSelectionChanged] #books=1
11:09:23.316 0001 [ConvertDGVControl.fillMetaInfoGaps] >>> 
11:09:23.316 0001 [ConvertDGVControl.fillMetaInfoGaps] <<< 
11:09:23.352 0001 [ConvertDGVControl.bookLibForm_BookDownloadSelectionChanged] #conv=1
11:09:23.403 0001 [ConvertDGVControl.snapshotSelection] upd sel rows=False
11:09:23.404 0001 [ConvertDGVControl.snapshotSelection] #conv sel=1
11:22:21.999 0001 [MainForm.OnClosing] >>> 
11:22:22.011 0001 [MainForm.OnClosing] <<< 
//

**Comment by audiamus** [comment]
> Using Windows Select All (Ctr+A) and Copy (Ctr+C) on the formatted catalog display then Paste (Ctr+V) that into Excel would save the list in a Excel spreadsheet. Would saving that copied text file afford any speed improvement versus traversing the SQLite db?

It's a bit more complex here. The big list is fed by a so-called data source which connects every single cell. And that's an interactive process. You'll notice some of that if you scroll the list and the bottom pane is filled additional meta data, the summary and the cover image. In AAX Audio Converter the list is just a list and the only connection to the associated audio file is the row index.

**Reply by Sailfish** [reply]
Yes, I see that now and, yes, I agree that the additional meta data is important to maintain for any subsequent reconstruction. It would be nice for me if there were a way to bypass re-traversing the SQLite db as an option. However, I realize that my large number of titles puts me in an outlier category and may not be an issue for the vast number of other BLC users.

I appreciate you looking into it and taking this into consideration.
